pinery-miso/src/main/java/ca/on/oicr/pinery/lims/miso/MisoClient.java

```diff
@@ -0,0 +1,1031 @@
+  }
+
+  @Override
+  public List<Run> getRuns() {
+    List<Run> runs = template.query(queryAllRuns, runMapper);
+    List<MisoRunPosition> positions = getRunPositions();
+    Map<Integer, Run> map = new HashMap<>();
+    for (Run r : runs) {
+      map.put(r.getId(), r);
+    }
+    for (MisoRunPosition p : positions) {
+      Run r = map.get(p.getRunId());
+      if (r != null) {
+        Set<RunPosition> rp = r.getSamples();
+        if (rp == null) {
+          rp = new HashSet<>();
+          r.setSample(rp);
+        }
+        rp.add(p);
+      }
+    }
+    return runs;
+  }
+
+  @Override
+  public Run getRun(Integer id) {
+    return getSingleRun(queryRunById, new Object[] { id });
+  }
+
+  @Override
+  public Run getRun(String runName) {
+    return getSingleRun(queryRunByName, new Object[] { runName });
+  }
+
+  private Run getSingleRun(String query, Object[] params) {
+    List<Run> runs = template.query(query, params, runMapper);
+    if (runs.size() != 1) return null;
+    Run run = runs.get(0);
+    Set<RunPosition> rp = new HashSet<>();
+    rp.addAll(getRunPositions(run.getId()));
+    run.setSample(rp);
+    return run;
+  }
+
+  private List<MisoRunPosition> getRunPositions() {
+    List<MisoRunPosition> positions = template.query(queryAllRunPositions, runPositionMapper);
+    List<MisoRunSample> samples = getRunSamples();
+    return mapSamplesToPositions(positions, samples);
+  }
+
+  private List<MisoRunPosition> getRunPositions(Integer runId) {
+    List<MisoRunPosition> positions = template.query(queryRunPositionsByRunId, new Object[] { runId }, runPositionMapper);
+    List<MisoRunSample> samples = getRunSamples(runId);
+    return mapSamplesToPositions(positions, samples);
+  }
+
+  private List<MisoRunPosition> mapSamplesToPositions(List<MisoRunPosition> positions, List<MisoRunSample> samples) {
+    Map<Integer, List<MisoRunPosition>> map = new HashMap<>();
+    for (MisoRunPosition p : positions) {
+      if (!map.containsKey(p.getPartitionId())) {
+        map.put(p.getPartitionId(), new ArrayList<>());
+      }
+      map.get(p.getPartitionId()).add(p);
+    }
+    for (MisoRunSample s : samples) {
+      List<MisoRunPosition> ps = map.get(s.getPartitionId());
+      if (ps != null) {
+        for (MisoRunPosition p : ps) {
+          Set<RunSample> rs = p.getRunSample();
+          if (rs == null) {
+            rs = new HashSet<>();
+            p.setRunSample(rs);
+          }
+          rs.add(s);
+        }
+      }
+    }
+    return positions;
+  }
+
+  private List<MisoRunSample> getRunSamples() {
+    return template.query(queryAllRunSamples, runSampleMapper);
+  }
+
+  private List<MisoRunSample> getRunSamples(Integer runId) {
+    return template.query(queryRunSamplesByRunId, new Object[] { runId }, runSampleMapper);
+  }
+
+  @Override
+  public List<Type> getTypes() {
+    return template.query(queryAllSampleTypes, typeMapper);
+  }
+
+  @Override
+  public List<AttributeName> getAttributeNames() {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)